tests/transformers/test_transformers_model_export.py

```diff
@@ -3798,3 +3799,82 @@ def predict(self, context, model_input, params=None):
+        "layer3": mock.Mock(type="disk"),
+        "layer4": mock.Mock(type="cpu"),
+    }
+
+    # Assert that the function returns True for a distributed model
+    assert _is_model_distributed_in_memory(mock_model) is True
+
+
+def test_model_on_single_device():
+    # Mocking a transformers_model with all components on cpu
+    mock_model = mock.Mock()
+    mock_model.device.type = "cpu"
+    mock_model.hf_device_map = {
+        "layer1": mock.Mock(type="cpu"),
+        "layer2": mock.Mock(type="cpu"),
+        "layer3": mock.Mock(type="cpu"),
+        "layer4": mock.Mock(type="cpu"),
+    }
+
+    # Assert that the function returns False for a non-distributed model
+    assert _is_model_distributed_in_memory(mock_model) is False
+
+
+class MockPipeline(transformers.Pipeline):
+    def _sanitize_parameters(self, *args, **kwargs):
+        # This is used in a Pipeline validation, so iterable collections must be returned
+        return {}, {}, {}
+
+    def preprocess(self, *args, **kwargs):
+        pass
+
+    def _forward(self, *args, **kwargs):
+        pass
+
+    def postprocess(self, *args, **kwargs):
+        pass
+
+
+def test_save_model_raises_exception_for_distributed_model(tmp_path):
+    mock_model = mock.Mock()
+    mock_model.device.type = "cpu"
+    mock_model.name_or_path = "dummy_name_or_path"
+    mock_model.hf_device_map = {
+        "layer1": mock.Mock(type="gpu", index=0),
+        "layer2": mock.Mock(type="gpu", index=1),
+        "layer3": mock.Mock(type="disk"),
+        "layer4": mock.Mock(type="cpu"),
+    }
+    mock_model.config.task_specific_params = {}
+
+    # NB: In order to avoid loading a model and filling up the disk in CI with some
+    # split configuration, this mock stack is being used to simulate a heterogenous
+    # memory setup for a mocked model.
+    # Due to the validations that occur within transformers when constructing a
+    # pipeline instance, a number of the validation stages need to be patched.
+    with mock.patch("transformers.utils.generic.infer_framework", return_value="pt"), mock.patch(
+        "transformers.pipelines.base.infer_framework_load_model", return_value=("pt", mock_model)
+    ), mock.patch("transformers.pipelines.base.is_torch_available", return_value=True):
+        mock_pipeline = MockPipeline(model=mock_model, tokenizer=None, device=torch.device("cpu"))
+
+        with mock.patch("mlflow.transformers._validate_transformers_model_dict"), mock.patch(
+            "mlflow.transformers._TransformersModel.from_dict", return_value=mock_pipeline
+        ), mock.patch("mlflow.transformers._is_model_distributed_in_memory", return_value=True):
+            with pytest.raises(
+                MlflowException,
+                match="The model that is attempting to be saved has been loaded into memory",
+            ):
+                mlflow.transformers.save_model(
+                    transformers_model=mock_pipeline, path=str(tmp_path / "model")
+                )
```

Comment:
This seems super fragile.

Reply:
yeah, it doesn't pass the different version checks at all. I moved to a simpler approach that actually uses accelerate basic device to module mappings on a tiny model to verify the behavior that we want here. 

Note: The "disk" mapping in the the test `test_basic_model_with_accelerate_device_mapping_fails_save` won't actually write to disk (provided the test logic works as intended), so there isn't any risk here of filling up the tiny disk storage on the CI runners during the test.